[ML] adding new KS test pipeline aggregation (#73334)
elastic/elasticsearch: docs/reference/aggregations/pipeline/bucket-count-ks-test-aggregation.asciidoc

The valid values are: "greater", "less", "two_sided". This parameter is key for
determining the K-S statistic used when calculating the K-S test.

`fractions`::
(Optional, list)
A list of doubles indicating the distribution of the samples with which to compare to the
`buckets_path` results. The default is a uniform distribution of the same length as the
`buckets_path` buckets.

`sampling_method`::
(Optional, string)
Indicates the sampling methodology when calculating the K-S test. Note, this is sampling
of the returned values. This determines the cumulative distribution function (cdf) points
used comparing the two samples. Default is `upper_tail`, which emphasizes the upper
end of the CDF points. Valid options are: `upper_tail`, `uniform`, and `lower_tail`.

==== Syntax

A `bucket_count_ks_test` aggregation looks like this in isolation:

[source,js]
--------------------------------------------------
{
  "bucket_count_ks_test": {
    "buckets_path": "range_values>_count", <1>
    "alternative": ["less", "greater", "two_sided"], <2>
    "sampling_method": "upper_tail" <3>
  }
}
--------------------------------------------------
// NOTCONSOLE

> <1> The buckets containing the values to test against.

Reviewer: ```suggestion
<1> The buckets containing the values to test against
```

Points 2 & 3 don't end in a full stop

{
  "aggs": {
    "buckets": {
      "terms": { <1>
        "field": "version",
        "size": 2
      },
      "aggs": {
        "latency_ranges": {
          "range": { <2>
            "field": "latency",
            "ranges": [
              { "to": 0.0 },
              { "from": 0, "to": 105 },
              { "from": 105, "to": 225 },
              { "from": 225, "to": 445 },
              { "from": 445, "to": 665 },
              { "from": 665, "to": 885 },
              { "from": 885, "to": 1115 },
              { "from": 1115, "to": 1335 },
              { "from": 1335, "to": 1555 },
              { "from": 1555, "to": 1775 },
              { "from": 1775 }
            ]
          }
        },
        "ks_test": { <3>
          "bucket_count_ks_test": {
            "buckets_path": "latency_ranges>_count",
            "alternative": ["less", "greater", "two_sided"]
          }
        }
      }
    }
  }
}
-------------------------------------------------
// TEST[setup:correlate_latency]

<1> The term buckets containing a range aggregation and the bucket correlation aggregation. Both are utilized to calculate
    the correlation of the term values with the latency.

> <2> The range aggregation on the latency field. The ranges were created referencing the percentiles of the latency field.

Reviewer: It would be nice to have a full example somewhere in docs of showing how the ranges were found

Author: Possibly, I think though that is a larger "how correlations + K-S tests work" type of docs.

Reviewer: Actually this isn't so bad as you explain that the latency ranges were calculated using a percentiles agg

> <3> The bucket count K-S test aggregation that determines if the count samples are from the same distribution as the uniform

Author: technically, no. I will think of a better way of phrasing it. 

The two-sampled K-S test (which we are doing here) is verifying that two samples come from the same distribution. In this particular case, we are testing if one sample (a uniform fraction sample) and another (these _counts) both come from the same distribution.

[role="xpack"]
[testenv="basic"]
[[search-aggregations-bucket-count-ks-test-aggregation]]
=== Bucket count K-S test correlation aggregation
++++
<titleabbrev>Bucket count K-S test aggregation</titleabbrev>
++++

experimental::[]

A sibling pipeline aggregation which executes a two sample Kolmogorov–Smirnov test
(referred to as a "K-S test" from now on) against a provided distribution and

> the distribution of documents counts in the configured sibling aggregation.

Reviewer 2: (I know you have some discussion of this later on, but I think this is crucial to understanding how this is meant to work for someone who knows what the K-S test actually does so worth including upfront.)

This test is useful to determine if two samples (represented by `fractions` and `buckets_path`) are
drawn from the same distribution.

[[bucket-count-ks-test-agg-syntax]]
==== Parameters

`buckets_path`::
(Required, string)
Path to the buckets that contain one set of values to correlate. Must be a `_count` path
For syntax, see <<buckets-path-syntax>>.

`alternative`::
(Optional, list)
A list of string values indicating which K-S test alternative to calculate.
The valid values are: "greater", "less", "two_sided". This parameter is key for
determining the K-S statistic used when calculating the K-S test. Default value is
all possible alternative hypothesis.

`fractions`::
(Optional, list)
A list of doubles indicating the distribution of the samples with which to compare to the
`buckets_path` results. The default is a uniform distribution of the same length as the
`buckets_path` buckets.

`sampling_method`::
(Optional, string)
Indicates the sampling methodology when calculating the K-S test. Note, this is sampling

> of the returned values. This determines the cumulative distribution function (cdf) points

Reviewer: ```suggestion
of the returned values. This determines the cumulative distribution function (CDF) points
```

Capitalised `CDF` is used in the next sentence

used comparing the two samples. Default is `upper_tail`, which emphasizes the upper
end of the CDF points. Valid options are: `upper_tail`, `uniform`, and `lower_tail`.

==== Syntax

A `bucket_count_ks_test` aggregation looks like this in isolation:

[source,js]
--------------------------------------------------
{
  "bucket_count_ks_test": {
    "buckets_path": "range_values>_count", <1>
    "alternative": ["less", "greater", "two_sided"], <2>
    "sampling_method": "upper_tail" <3>
  }
}
--------------------------------------------------
// NOTCONSOLE
<1> The buckets containing the values to test against.
<2> The alternatives to calculate.
<3> The sampling method for the K-S statistic.

[[bucket-count-ks-test-agg-example]]
==== Example

The following snippet runs the `bucket_count_ks_test` on the individual terms in the field `version` against a uniform distribution.
The uniform distribution reflects the `latency` percentile buckets. Not shown is the pre-calculation of the `latency` indicator values,
which was done utilizing the
<<search-aggregations-metrics-percentile-aggregation,percentiles>> aggregation.

> This example is only using the 10s percentiles.

Reviewer 2: ```suggestion
This example is only using the deciles of `latency`.
```

                "to" : 665.0,
                "doc_count" : 20
              },
              {
                "key" : "665.0-885.0",
                "from" : 665.0,
                "to" : 885.0,
                "doc_count" : 20
              },
              {
                "key" : "885.0-1115.0",
                "from" : 885.0,
                "to" : 1115.0,
                "doc_count" : 10
              },
              {
                "key" : "1115.0-1335.0",
                "from" : 1115.0,
                "to" : 1335.0,
                "doc_count" : 0
              },
              {
                "key" : "1335.0-1555.0",
                "from" : 1335.0,
                "to" : 1555.0,
                "doc_count" : 0
              },
              {
                "key" : "1555.0-1775.0",
                "from" : 1555.0,
                "to" : 1775.0,
                "doc_count" : 0
              },
              {
                "key" : "1775.0-*",
                "from" : 1775.0,
                "doc_count" : 0
              }
            ]
          },
          "ks_test" : {
            "less" : 0.9642895789647244,
            "greater" : 4.58718174664754E-9,

>             "two_sided" : 4.58718174664754E-9

Reviewer 2: I would expect `greater` to be very nearly `two_sided` / 2 in this case.

Author: Yeah, the `two_sided` isn't accurate. I think Hodge's approximation only works in the sided case. We will probably have to do something else.
